### PR TITLE
fix(mme): [AMF: Unit Test] Creating ue_context infrastructure#8347

### PR DESCRIPTION
**Pull request by Krishnavamsi-wavelabs**
Signed-off-by: krishnavamsi-wavelabs <krishna.vamsi@wavelabs.ai>

## Summary
Added test_fixtures for common code
For instance in setup :
a. Created amf_create_new_ue_context()
In tear Down
b. Cleaned up the variables.

## Test Plan

Verified basic sanity with UERANSIM.
Please find attached logs and pcap snap for basic registration and PDU Session setup

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by Krishnavamsi-wavelabs** [comment]
snapshot of ncap:
Screenshot (43)

test cases compiled:
Screenshot (30)

logs:
[mme_warning_cleanup.log](https://github.com/magma/magma/files/7204562/mme_warning_cleanup.log)

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp [hunk 1]
```diff
@@ -95,17 +95,24 @@ uint8_t NAS5GPktSnapShot::registration_reject[4] = {0x00, 0x00, 0x00, 0x00};
...
+  NAS5GPktSnapShot* nas5g_pkt_snap;
   RegistrationRequestMsg reg_request;
-  bool decode_res = false;
-
-  uint32_t len = nas5g_pkt_snap.get_reg_req_buffer_len();
+  bool decode_res;
+  virtual void SetUp() {
+    nas5g_pkt_snap = new NAS5GPktSnapShot();
+    decode_res     = false;
+    memset(&reg_request, 0, sizeof(RegistrationRequestMsg));
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf.cpp]: initialize using `RegistrationRequestMsg reg_request = {0};`

Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp]: initializing with RegistrationRequestMsg reg_request = {0};  will give the following error.
error

so, initialized as RegistrationRequestMsg reg_request = {}
tests

Comment (ardzoht) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp]: @Krishnavamsi-wavelabs why not initialize with `new` constructor? the same way you're initializing `NAS5GPktSnapShot`

**Review by ulaskozat — APPROVED** [review]
LGTM

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp [hunk 2]
```diff
@@ -402,26 +402,27 @@ TEST(test_amf_nas5g_pkt_process, test_amf_service_accept) {
...
+ protected:
+  ue_m5gmm_context_s* ue_context;
+
+  virtual void SetUp() { ue_context = amf_create_new_ue_context(); }
+  virtual void TearDown() { delete ue_context; }
+};
+
+TEST_F(AmfUeContextTest, test_ue_context_creation) {
   EXPECT_TRUE(nullptr != ue_context);
   EXPECT_TRUE(0 == ue_context->amf_teid_n11);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf.cpp:436]: why did you move the tested function call to SetUp() ? This does not make much sense to me.

**Review by ulaskozat — CHANGES_REQUESTED** [review]
See my comment.

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp (hunk 1 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp]: @ardzoht  ,
addressed review comment.
both NAS5GPktSnapShot,RegistrationRequestMsg structures defined normal member variables, dynamic memory allocation not needed.

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp (hunk 2 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp:436]: @ulaskozat,
we would like to introduce text fixtures to move reusable code.
and existing multiple test cases having ue_context allocation and deallocation.

Please provide your feedback if the improvement is needed here.

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp (hunk 1 above)
Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp]: Screenshot (51)

**Comment by Krishnavamsi-wavelabs** [comment]
Screenshot (51)

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp (hunk 2 above)
Comment (ulaskozat) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp:436]: Hmm. It is a bit odd that `test_ue_context_creation` is not actually calling any function. Is there a plan to add more test cases here?

Comment (Krishnavamsi-wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp:436]: agree.
we have seen a couple of memory corruptions due to uninitialized variables.
At that time test_ue_context_creation test case was added to check uninitialized members.

yes, we had a plan to add more test cases.
we are working on increasing Unittesting coverage as per the plan.

Comment (ulaskozat) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp:436]: Okay let's leave it as is for now and we can reassess later based on the new test cases being added.

**Review by ulaskozat — APPROVED** [review]
Let's merge this.
